main.cpp

```diff
@@ -30,28 +33,234 @@ Project 5: Part 1 / 4
  Wait for my code review.
  */
 
+//Required UDT for Copied UDT 1 to work
+
+struct BoulderProblem
+{
+    int problemGrade;
+    double wallAngle;
+    
+    BoulderProblem() : problemGrade( 3 ), wallAngle( 30 ) {}
+    ~BoulderProblem() 
+    {
+        std::cout << "BoulderProblem dstor\n";
+    }
+
+    struct Hold
+    {
+        int holdType { 2 };
+        double holdSize; 
+        double holdHeight;
+
+        Hold() : holdSize( 0.8 ), holdHeight( 1 ) {}
+        ~Hold() 
+        {
+            std::cout << "Hold dstor\n";
+        }
+
+        void printHoldInfo();
+    };
+
+    double calculateDifficulty( double ropeLength );
+
+    Hold crimp;
+};
+
+void BoulderProblem::Hold::printHoldInfo()
+{
+    std::cout << "Hold Type: " << holdType;
+    std::cout << "\nHold Size: " << holdSize;
+    std::cout << "\nHold Height: " << holdHeight;
+    std::cout << std::endl;
+}
+
+double BoulderProblem::calculateDifficulty( double ropeLength )
+{
+    std::cout << "Calculating difficulty for V" << problemGrade << " problem at a " << wallAngle << " degree incline.\n";
+    return ((problemGrade * crimp.holdType) / crimp.holdSize) - ropeLength;
+}
+
 /*
  copied UDT 1:
  */
 
+struct TopRopeRoute
+{
+    double wallAngle { 10 };
+
+    TopRopeRoute(){}
+    ~TopRopeRoute()
+    {
+        std::cout << "TopRopeRoute dstor\n";
+    }
+
+    struct RouteGrade
+    {
+        double gradeNumber;
+        char gradeLetter; 
+        
+        RouteGrade() : 
+        gradeNumber( 5.10 ), 
+        gradeLetter( 100 ) // ascii letter 'd' 
+        {}
+
+        ~RouteGrade()
+        {
+            std::cout << "RouteGrade dstor\n";
+        }
+
+        void printGradeInfo();
+    };
+
+    void buildRoute( int moves, double wallHeight );
+
+    RouteGrade hard;
+};
+
+void TopRopeRoute::RouteGrade::printGradeInfo()
+{
+    std::cout << "Grade is: " << gradeNumber << gradeLetter << std::endl;
+}
+
+void TopRopeRoute::buildRoute( int moves, double wallHeight = 40.36 )
+{
+    BoulderProblem::Hold hold;
+
+    std::cout << "The wall is at a " << wallAngle << " degree incline.\n";
+
+    for ( int i = moves; i > 0; i-- )  // already have a for loop
+    {
+         hold.holdHeight = ( wallHeight / moves ) * i;
+         std::cout << "Hold height for move " << i << " is: " << hold.holdHeight << std::endl;
+    }
+}
+
 /*
  copied UDT 2:
  */
 
+struct Mountain 
+{
+    int height { 15 };
+    int routes { 20 };
+
+    Mountain(){}
+    ~Mountain()
+    {
+        std::cout << "Mountain dstor\n";
+    }
+
+    void mountainFeatures( TopRopeRoute face, BoulderProblem base, double mountain );
+
+    void constructMountain( double baseDiameter );
+};
+
+void Mountain::mountainFeatures(TopRopeRoute face, BoulderProblem base, double mountain )
+{
+    std::cout << "Mountain height is " << height << " feet and has " << routes << " routes\n";
+
+    for ( int i = this->routes; i > 0; i-- ) // already have a for loop
```

Comment:
You should use pre-decrement, not post-decrement: `--i`
- [ ] use post-decrement, not pre-decrement